README.md
tmnw
====
The simple network loop.


## Summary
tmnw is a two file library for managing network messages using non-blocking sockets.

Libraries like libevent and libuv will outperform tmnw on raw speed, but tmnw aims to be 
simpler to use than libevent or libuv by adhering to some simple goals.

<ol>
<li>Be useful with two files (tmnw.c and tmnw.h), one object (tmnw.o)</li>
<li>Operate well in both single threaded and multi threaded environments</li>
<li>Do not try to run on every Unix known to man</li>
<li>Allow an implementor to use callbacks to handle protocols that need to run on top of these platforms (see Usage for more details)</li>
<li>Focus on the network (UDP and TCP, not really worried about IPC)</li>
</ol>


## Dependencies
tmnw has only one real dependency: poll().  It can be found on most Unix systems and will give you a nice level of performance right out of the box.  See `man poll` for more information.  Windows is another story.


## Usage

Using tmnw in your application is pretty simple. Unfortunately, I only 
have directions for C right now, as I have not had time to test with C++.


### Compiling

To compile `tmnw` for your own apps, do the following:

<p>
<h3>Linux</h3>
<code>gcc -Wall -std=c99</code> should yield the library you need.  If you do run into any build errors, please notify me at ramar.collins@gmail.com or send me a pull request.
</p>

<p>
<h3>OSX</h3>
tmnw is so far untested on OSX, but it is down the pipeline as I have a Mac Mini to test with.
</p>

<p>
<h3>Windows</h3>
Windows has poll() support now, but internet lore tells me this was not always the case.  As time avails, I may look into bringing this to Windows as it could be useful for game developers.
</p>



### Example

Below you can see some example code on how to actually use `tmnw` as a networking backend.
The following example shows you how to set up a TCP listener on port 2000.

<ol>
<li>
Declare and define a Socket.

<p>
So, whenever the term Socket is read in this manual, realize that I'm not speaking about an actual file that you can send and receive data packets over.   In this library, the term 'Socket' refers to a datatype that serves as a template for a connection you would like to initiate.  For example, you can tell the library that you intend to make a server that only listens for UDP messages from a certain port.  The C code for it is below:
</p>

<pre>
	Socket sk   = { 
		.server   = 1,          // 'True' means we want a server, 'False' makes a client
		.proto    = "tcp",      // Select TCP, not UDP
		.port     = 2000,       // Choose a port from 1 to 65534
		.hostname = "localhost" // Choose a hostname
	};
</pre>
</li>

<li>
Open a socket and bind and listen if successful

<p>
In Unix-like systems, creating a socket is, frankly, a pain in the ass.  There are tons of flags and behaviors that you need to keep track of that really won't matter to you if you're not planning doing anything particularly unusual with the connection.
</p>

<p>
Also notice that you <b>don't</b> need to check errno when using this library.  Each of the functions was written to return either true or false depending on whether or not it was successful.
</p>

<pre>
	//Notice how we can chain these together
	if ( !socket_open(&sock) || !socket_bind(&sock) || !socket_listen(&sock) )
	{
		fprintf( stderr, "Couldn't create socket.\n" );
		return 0;
	}
</pre>

<p>
NOTE: WebSockets and other bi-directional communication schemes would count as unusual in my opinion.  These protocols really <i>aren't</i> why `tmnw` was written.  So if that's your need, this is probably not the library you're looking for.
</p>

</li>


<li>
Check and initialize the socket structure

<p>
Unfortunately, we are still writing in C and zero'ing out and initializing data is still a task I have to delegate to you.  Fortunately, this simple function makes it really easy for you to do.  This function also handles any validation issues that you may have in your code, such as invalid ports or plain errant setup.
</p>
 
<pre>
	if (!initialize_selector(&sel, &sock))
		return tmnw_err(&sel);
</pre>
</li>

<li>
Start a non-blocking server loop

<p>
`tmnw` does not block at all.  It will never set up a blocking server no matter what you do.  Yay. 
</p>

<pre>
	if (!activate_selector(&sel))
		return tmnw_err(&sel);
</pre>
</li>


<li> 

Test that all is well.
<p>
	If you've gotten this far and have compiled the program, you'll find that it does nothing but tell you which port it is running on.   But hey, my friend, you'd be mistaken.  Because there is an invisible server running via port 2000 that is just waiting on you to send it something...
</p>

<p>
	You can run the following script via Bash to see how it works.
	You can even test that those non-blocking sockets are actually non-blocking working as they should.
</p>

<pre>	
	#/bin/sh -
	for n in `seq 1 10`; do
		curl http://localhost:2000 &
	done
</pre>

<p>
	The above script just asks curl to make a GET request for "/" 10 times.
</li>
</ol>


## Doing Slightly More Than the Minimum

<p>
So far, our server is not doing very much except sending back a piddly little message and calling it a day.
Right, so you just wasted all of your time reading through this manual.  Good for you, you've won the "Captain Dingus" award.
</p>

<p>
Fortunately, there is more to the story.  tmnw tries pretty hard to abstract the process of network communication.  To make it easy to write apps that can handle networking, tmnw breaks up the server loop into four distinct parts.  This way, you don't even have to write your own event loop.  You are just writing handlers that plug in to said loop at different parts of the process.
</p>


## Handlers

<p>
	Now we can define "actions" for our application to take at certain points
	in the request-response chain.  
</p>

<p>
	In the theoretical sense, a handler is really just a specific action to follow when a certain condition is met.  For example, let's suppose you want to do more than send my trite little message back to your users.   Perhaps you want to show a picture.  Or maybe you want to be a real web server and handle requests for certain directories on your system.  Or maybe you don't want anybody looking at anything ever, so you decide to automatically '400 Unauthorized' headers in succession.   Perhaps you can serve rudimentary CSS along with this message so that your users can '400 Unauthorized' headers with pages in colors matching each spectrum of the rainbow.   Whatever floats your boat, my friend.
</p>

<p>
	In the technical sense, a handler is nothing more than a function pointer.  
	When writing your own handlers, use the following defintion. 
	The following snippet shows the declaration you'll need to use when binding custom actions to `tmnw`.
<p>

<pre>
//Always return a \_Bool, works via both C and C++
\_Bool (\*fn_name) (Recvr \*r, void \*ud, char \*err) 
</pre>

<p>
Looks pretty easy, right?  Except, what's that Recvr * thing?   Well, I'm glad you asked.
</p>

<p>
The Recvr structure is used by tmnw to keep track of the status of each request.  To keep it simple, pretty much <i>every</i> possible piece of metadata about an open socket connection is stored here.   I could go into great detail about how much data is stored here, but I'll let the code speak for me this time:
</p>

<pre>
typedef struct {
  Socket            child;  //Child file descriptor
	Stage             stage;  //An enum defining where we are currently in the process
	int              rb, sb;  //Bytes received at an invocation of the "read loop"
  int         recvd, sent;  //Total bytes sent or received
  int                 len;  //Length of message to be sent?
	int               error;  //Was an error detected?

	//For error messages when everything may fail
  uint8_t          errbuf[NW_ERROR_BUFFER_LENGTH];  

	//Notice that tmnw can set a size limit on received requests.
#ifdef NW_BUFF_FIXED
  uint8_t        request_[NW_MAX_BUFFER_SIZE];
  uint8_t       response_[NW_MAX_BUFFER_SIZE];
#endif
  uint8_t        *request;
  uint8_t       *response;
	
	//What the client sends goes here
	Buffer       _request;

	//What you prepare goes here
	Buffer      _response;

	//...you can skip these...
	int      sretry, rretry;
	int          recv_retry;
	int          send_retry;
	int          *socket_fd;  //Pointer to parent socket?

	//The original pollfd structure
	struct pollfd   *client;  //Pointer to currently being served client

#ifndef NW_DISABLE_LOCAL_USERDATA
	void          *userdata;
#endif

	//Track the time that a request starts and ends, also 
	//used to track how long the connection has been opened.
	struct timespec start;
	struct timespec end;
} Recvr;
</pre>

<li>
NW_AT_READ
<p>
The stage immediately after tmnw reads a message from a client
</p>
</li>

<li>
NW_AT_PROC
<p>An optional intermediate stage that tmnw reaches after successfully reading a  client's message.  Assumes that no further reading is needed.</p>
</li>

<li>
NW_AT_WRITE
<p>
Stage in which message has been sent from server.  If the message were extremely large, one could optionally send the rest of the message to the client from this stage.
</p>
</li>

<li>
NW_AT_COMPLETED
<p>
The message is assumed to be done at this point.  You can close the file descriptor or keep it open if you are using some kind of protocol that is capable of re-using an open socket.
</p>
</li>



Customization
=============
The	list of defines to change the size of tmnw are below:

	NW_BEATDOWN_TEST - Enables some test friendly options, like stopping after
										 a certain number of requests
	NW_CATCH_SIGNAL  - Catch signals or not.  Will define free_selector by default.
	NW_VERBOSE       - Be verbose
	NW_FOLLOW        - Dump a message when moving through each part of the program.
	NW_SKIP_PROC     - Compile without a "processing" part of the server loop.
	NW_KEEP_ALIVE    - Define whether or not the connection should stay open after 
										 reaching the write portion of the connection's life cycle.
	NW_LOCAL_USERDAT - Define whether or not to use local userdata.
	NW_GLOBAL_USERDA - Define whether or not to use global userdata.
	NW_STREAM_TYPE   - Define which way to read requests...


Here is an example of using a header file 
to define settings. We'll call our file
whatever.h:

<pre>
//Contents of "whatever.h"
#define MIN_ACCEPTABLE_READ 64   	//Kick requests that fail to read this much
#define NW_MAX_BUFFER_SIZ   8192 	//Read no more than this much to buffer
#define NW_MAX_BUFFER_SIZ   8192 	//Write no more than this much to buffer
#define MAX_OPEN_FILES      256  	//How many files can I have open at once?
#define READ_TO_FILE        0    	//When reading, use a file and stream
#define READ_TO_BUF         1    	//When reading, use a buffer
#	-DNW_FOLLOW     - Print each branchable call as the program passes 
#    through execution.
#	-DNW_VERBOSE    - Print each error message to stderr if anything fails.
#	-DNW_KEEP_ALIVE - Define whether or not to leave connections open 
#    after NW_COMPLETED is reached.
#	-DNW_SKIP_PROC  - Define whether or not to skip the processing step 
#    of a connection. 
# -DNW_MIN_ACCEPTABLE_READ   - Define a floor for the number of bytes 
#    received during read and drop the connection if it's not reached.
# -DNW_MIN_ACCEPTABLE_WRITE  - Define a floor for the number of bytes 
#    received during write and drop the connection if it's not reached.
#	-DNW_LOCAL_USERDATA  - Compile with space for local userdata.
#	-DNW_GLOBAL_USERDATA - Compile with space for global userdata.
#	-DNW_CATCH_SIGNAL    - Listen for signals...
</pre>


CHANGELOG
Wed Jan 10 19:30:20 EST 2018
	 -After a very long hiatus dealing with networking, I've finally come back to make it work without excess dependencies.
	 -The solution was to extract the needed functionality from single.[ch] and build it into the library.  A better way to handle this would be to use native networking code.

Sat Feb  4 04:03:01 EST 2017
	 -Updated nw.h to make defines a little clearer.
	 -Reorganized nw.c to make all code a bit cleaner.
	 -Added error indicator to Recvr structure.  This enables nw to send a message using either the currently chosen buffer or the fixed buffer called 'error'.

Fri Jan 27 17:18:21 EST 2017
	 -Tested POSTs of 100KB+.  Will try larger tests and more memory.  This works great!  150 at a time worked smoothly.

Fri Jan 27 12:49:38 EST 2017
	 -Added some new #define flags.  NW_QUEUE_* will allow nw to hold as many requests in queue as possible.  This is only for testing, but its been instrumental in getting to the next steps.
	 -Tested with 1000 108 byte GET requests.  Haven't seen a crash or memory mismatch.  Also compiled with clang's super strict warnings and haven't had any real issues.
	 -Fixed buffer library.  On bf_free(), size, written and error were not being reset to zero.  Caused some problems but we're good now.
	 -NW_QUEUE_WRITES needs some work.

Fri Jan 27 00:17:30 EST 2017
	 -Fixed bug in main loop.  Was initializing poll structures out of order.  
	 -This caused malloc'd memory for responses and requests to point to the wrong structure. (so fd 3's memory might really be fd 4's memory and in some cases fd 4's memory wasn't initialized)

Sat Jan 21 05:03:20 EST 2017
	 -A lot of this log is missing.  I will go back and find those notes....
	 -Switching most of the backend to start using nw_add, nw_read and nw_write.  These serve as bounds checking functions to add to buffers.

Sun Oct 16 12:16:40 EDT 2016
	 -Fixed a few errors in nw.  Was sending WAY too much data when trying to ship packets via write().
	 -Added reset_recvr() to ameliorate tearing down sockets.   It does not close the file descriptor, it just wipes buffers and resets any counts.
	 -Moved around data structures and defines to make the code easier to parse.a
	 -Added reset_recvr() before handlers on all error codes (I'm probably going to get rid of this,  it makes little sense...)
	 -(specifically because the handler should be in control of ALL of that)
	 -Added an event handler for accept().  It's there, but has to be enabled.   This should help block bullshit requests before they even make it to the processing step.  

Thu Sep 29 17:17:51 EDT 2016
	 -Finished getting basic reads and writes to work.  Have not yet tried retries.
	 -Also noticed some things that a user may ALWAYS want to do: specifically, use memset or some other means to reset the Recvr structure.   If data is kept there, it will throw off subsequent reads and writes (and you'll probably get garbage data or worse, segfault)
	 -Added many useful macros, especially NW_CALL which let's me follow the progression of "important" code by printing it's position and the branch that is being executed. It can be def'd out by leaving -DNW_FOLLOW off of the gcc flags.
	 -I overlooked the nuances of the handle() macro.   I originally only handled failures with this macro, but this didn't work for read, write and process handlers.   uhandle handles callbacks for both failures and successes.  It also uses a different Executor table (please see nw.h for more info on this structure)
	 -Version up (0.2)

Thu Sep 29 02:15:22 EDT 2016
	 -Got through most of the server loop.  Running into some issues with pollfd strucutre references at _write in main.c

Sun Sep 25 14:38:51 EDT 2016
	 -I added a change, lookit me, ma!


nw.h
/* ------------------------------------- *
	nw.h
	----

 	The simple network loop.

	Summary
	-------
	nw is a lighteweight library using poll 
	to assist in quickly setting up servers
	capable of handling non blocking sockets.
	It is written in such a way that handler
	functions can be inserted like a game loop 	

	Usage
	-----
	Using nw is pretty simple as the backend
	to your server application is pretty 
	simple:

	1. open a socket 

	Defines
	-------
	NW_BEATDOWN_MODE - Enables some test friendly options, like stopping after
										 a certain number of requests
	NW_CATCH_SIGNAL  - Catch signals or not.  Will define free_selector by default.
	NW_VERBOSE       - Be verbose
	NW_FOLLOW        - Dump a message when moving through each part of the program.
	NW_SKIP_PROC     - Compile without a "processing" part of the server loop.
	NW_KEEP_ALIVE    - Define whether or not the connection should stay open after 
										 reaching the write portion of the connection's life cycle.
	NW_LOCAL_USERDAT - Define whether or not to use local userdata.
	NW_GLOBAL_USERDA - Define whether or not to use global userdata.
	NW_STREAM_TYPE   - Define which way to read requests...

	Customization
	-------------
	Here is an example of using a header file 
	to define settings. We'll call our file
	whatever.h:

  //Contents of "whatever.h"
 #define MIN_ACCEPTABLE_READ 64   	//Kick requests that fail to read this much
 #define NW_MAX_BUFFER_SIZ   8192 	//Read no more than this much to buffer
 #define NW_MAX_BUFFER_SIZ   8192 	//Write no more than this much to buffer
 #define MAX_OPEN_FILES      256  	//How many files can I have open at once?
 #define READ_TO_FILE        0    	//When reading, use a file and stream
 #define READ_TO_BUF         1    	//When reading, use a buffer

	Bugs
	----
	Please notify Antonio R. Collins II at
	ramar.collins@gmail.com.

	
 * ------------------------------------- */
#include <poll.h>
#include "buff.h"

#define LITE_BUFFER

#ifndef NW_H
#define NW_H

/*nw has a few sensible defines*/
#ifndef NW_ERROR_BUFFER_LENGTH
 #define NW_ERROR_BUFFER_LENGTH      4096 
#endif
#ifndef NW_MIN_ACCEPTABLE_READ
 #define NW_MIN_ACCEPTABLE_READ        32
#endif
#ifndef NW_MIN_ACCEPTABLE_WRITE
 #define NW_MIN_ACCEPTABLE_WRITE       32
#endif
#ifndef NW_MAX_ACCEPTABLE_READ
 #define NW_MAX_ACCEPTABLE_READ      4096
#endif
#ifndef NW_MAX_ACCEPTABLE_WRITE
 #define NW_MAX_ACCEPTABLE_WRITE     4096
#endif
#ifndef NW_RETRY_READ
 #define NW_RETRY_READ                  3
#endif
#ifndef NW_RETRY_WRITE
 #define NW_RETRY_WRITE                 3
#endif
#ifndef NW_MAX_EVENTS
 #define NW_MAX_EVENTS                100
#endif
#ifndef NW_MAX_BUFFER_SIZE
 #define NW_MAX_BUFFER_SIZE         64000 
#endif
#ifndef NW_STREAM_TYPE
 #define NW_STREAM_TYPE               "?"
#endif
#ifndef NW_BUFF_FIXED
 #define NW_BUFF_DYNAMIC
#endif
#if 0
#ifndef NW_DISABLE_LOCAL_USERDATA
 #define NW_DISABLE_LOCAL_USERDATA    0 
#endif
#ifndef NW_DISABLE_GLOBAL_USERDATA
 #define NW_DISABLE_GLOBAL_USERDATA   0 
#endif
#endif
//Choose whether or not to queue writes
#ifdef NW_QUEUE_WRITES
 #define NW_QUEUE_WRITE_DIRNAME "local"
#endif


//Follow program flow
#ifdef NW_FOLLOW
 #define NW_CALL(c) \
	(c) || (fprintf(stderr, "%s: %d - %s\n", __FILE__, __LINE__, #c)? 0: 0)
#else
 #define NW_CALL(c) \
	c
#endif

//Return error messages and a code at the same time.
#ifdef NW_VERBOSE
 #define nw_err(c, ...) \
	(fprintf(stderr, __VA_ARGS__) ? c : 0)
#else
 #define nw_err(c, ...) \
	c
#endif

//Reset read event
#define nw_reset_read() \
	r->client->events = POLLRDNORM

//Reset write event
#define nw_reset_write() \
	r->client->events = POLLWRNORM

//Reset the receiver
#define nw_reset_recvr() \
	memset(r, 0, sizeof(Recvr)); close((&r->client)->fd); (&r->client)->fd = -1

//Get fd without worrying about pollfd structure
#define nw_get_fd() \
	r->client->fd

/*nw's "static" defines, so I don't forget what goes where*/
enum { NW_RECV = 0, NW_SEND };

/*stream selections*/
typedef enum 
{
	NW_STREAM_BUF = 0, 
	NW_STREAM_FD,
	NW_STREAM_PIPE
} Stream;

/*Error map or a big struct?*/
enum {
	ERR_SO_HAPPY_THERE_ARE_NO_ERRORS = 0,

	/*Poll errors*/
	ERR_POLL_INITIAL_ALLOCATOR,
	ERR_POLL_TOO_MANY_FILES,
	ERR_POLL_RECVD_SIGNAL, 

	/*Socket child spawn*/	
	ERR_SPAWN_ACCEPT,
	ERR_SPAWN_NON_BLOCK_SET,
	ERR_SPAWN_MAX_CLIENTS,

	/*Accepted client select/poll loop*/
	ERR_READ_CONN_RESET,
	ERR_READ_EGAIN,
	ERR_READ_EBADF,
	ERR_READ_EFAULT,
	ERR_READ_EINVAL,
	ERR_READ_EINTR,
	ERR_READ_EISDIR,
	ERR_READ_CONN_CLOSED_BY_PEER,
	ERR_READ_BELOW_THRESHOLD,
	ERR_READ_MAX_READ_RETRY_REACHED,

	/*Processor errors go here*/
	ERR_WRITE_CONN_RESET,
	ERR_WRITE_EGAIN,
	ERR_OUT_OF_MEMORY,
	ERR_REQUEST_TOO_LARGE,
	ERR_WRITE_EBADF,
	ERR_WRITE_EFAULT,
	ERR_WRITE_EFBIG,
	ERR_WRITE_EDQUOT,
	ERR_WRITE_EINVAL,
	ERR_WRITE_EIO,
	ERR_WRITE_ENOSPC,
	ERR_WRITE_EINTR,
	ERR_WRITE_EPIPE,
	ERR_WRITE_EPERM,
	ERR_WRITE_EDESTADDREQ, /*For UDP*/
	ERR_WRITE_CONN_CLOSED_BY_PEER,
	ERR_WRITE_BELOW_THRESHOLD,
	ERR_WRITE_MAX_WRITE_RETRY_REACHED,

	/*Failure to allocate within the buffer*/
	/*ERR_BUFF_REALLOC_FAILURE,
	ERR_BUFF_OUT_OF_SPACE,*/
	ERR_TIMEOUT_CONN,
	ERR_END_OF_CHAIN,
};


//Datatype to track where nw is in the request->reply process
typedef enum 
{
	NW_AT_READ = 0,   /*The current connection is read ready.*/
	NW_AT_PROC,       /*The current connection is processing a response*/
	NW_AT_WRITE,      /*The current conenction is write ready*/
	NW_COMPLETED,     /*The current connection is done*/
	NW_AT_ACCEPT,     /*The current connection is about to be created*/
	NW_ERR            /*The current connection ran into some sort of error*/
} Stage;



//Datatype to manage buffering of receipt and relay of network messages
typedef struct {
  Socket            child;
	Stage             stage;
	int              rb, sb;  //Bytes received at an invocation of the "read loop"
  int         recvd, sent;  //Total bytes sent or received
  int                 len;  //Length of message to be sent?
	int               error;  //Was an error detected?
  uint8_t          errbuf[NW_ERROR_BUFFER_LENGTH];  //For error messages when everything may fail
#if 0
  int          request_fd;  
  int         response_fd;
#endif
#ifdef NW_BUFF_FIXED
  uint8_t        request_[NW_MAX_BUFFER_SIZE];
  uint8_t       response_[NW_MAX_BUFFER_SIZE];
#endif
  uint8_t        *request;
  uint8_t       *response;
	Buffer       _request;
	Buffer      _response;
	int      sretry, rretry;
	int          recv_retry;/*3*/
	int          send_retry;/*5*/
	int          *socket_fd;  //Pointer to parent socket?
	struct pollfd   *client;  //Pointer to currently being served client
#ifndef NW_DISABLE_LOCAL_USERDATA
	void          *userdata;
#endif
	//This allows nw to cut connections that have been on too long
	struct timespec start;
	struct timespec end;
} Recvr;


//Datatype to control event handlers
typedef struct { 
	_Bool (*exe)(Recvr *r, void *ud, char *err);
	enum {
		NW_CONTINUE = 0,
		NW_NOTHING,
		NW_RETURN,
		NW_EXIT
	}       action;
	int     status;
	char    err[2048]; 
} Executor;


//Dataype to control "stream" handlers
typedef struct {
	/*Should support writing to file or to memory*/ 
	//_Bool (*exe)(void *in, void *out);
	_Bool (*read) (Recvr *r);
	_Bool (*write)(Recvr *r);
} Streamer;


//Datatype to set up a server or client
typedef struct 
{
	int       read_min   ,  /*How much data needs to be read at a time?*/
	          write_min  ;  /*How much data needs to be written at a time?*/
	int8_t    recv_retry ,  /*Number of times to retry reading*/
	          send_retry ;  /*Number of times to retry sending*/
	Stream    stream     ;  /*Which stream to use*/
	Executor *errors,       /*Error handlers*/
           *runners    ;  /*Connection life cycle handlers*/
	int       run_limit  ;  /*A time limit to cut long running connections*/ 
	int       max_events ;  /*How many events should I allow to queue at a time?*/
	_Bool     keepalive  ;  /*Should the user be responsible for closing connections?*/

#ifndef NW_DISABLE_GLOBAL_USERDATA
	void     *global_ud  ;  /*Global userdata (not copied)*/
#endif

#ifndef NW_DISABLE_LOCAL_USERDATA
	int       lsize      ;  /*Size of one local userdata*/
	void    **local_ud   ;  /*Local userdata (this src doled out per max_events)*/
#endif

#ifdef NW_BEATDOWN_MODE
	int       stop_at    ;  /*Stop serving after x amount of requests*/
#endif

	//Stuff I'll never set
	struct    
   pollfd  *clients;
	int       tsize      ;  /*Total size of all local userdata (not touched)*/
	Socket   *parent     ;  /*The big daddy socket of whatever server you're running*/
	Recvr    *rarr       ;  /*Array to choose which "receiver" you want*/
} Selector;


//All forward declarations
extern Executor _nw_errors[ERR_END_OF_CHAIN + 1];
_Bool initialize_selector (Selector *, Socket *);
void free_selector (Selector *s);
_Bool activate_selector (Selector *s); 
void print_selector (Selector *s);
_Bool executor (Recvr *r, void *ud, char *err); 
_Bool nw_close_fd (Recvr *r, void *ud, char *err);
_Bool reset_read_fd (Recvr *r, void *ud, char *err);
_Bool reset_write_fd (Recvr *r, void *ud, char *err);
_Bool reset_buffer (Recvr *r, void *ud, char *err);
_Bool nw_close_fd (Recvr *r, void *ud, char *err);
_Bool nw_finish_fd (Recvr *r, void *ud, char *err);
_Bool nw_reset_fd (Recvr *r, void *ud, char *err);
_Bool nw_add_read (Recvr *r, uint8_t *msg, int size);
_Bool nw_add_write (Recvr *r, uint8_t *msg, int size); 
void nw_set_error (Recvr *r);
#endif
